src/dictionary.c

```diff
@@ -429,16 +518,24 @@ SEXP vec_match(SEXP needles, SEXP haystack) {
     }
   }
 
-  struct dictionary* d_needles = new_dictionary_partial(needles);
+  struct dictionary* d_needles = new_dictionary_params(needles, true, na_equal);
   PROTECT_DICT(d_needles, &nprot);
 
   // Locate needles
   SEXP out = PROTECT_N(Rf_allocVector(INTSXP, n_needle), &nprot);
   int* p_out = INTEGER(out);
 
+  bool propagate = !na_equal;
+
   for (int i = 0; i < n_needle; ++i) {
+    if (propagate && dict_is_missing(d_needles, i)) {
```

Comment:
Does it make sense to pull the `propagate` check out of the loop for performance?

```c
if (propagate) {
    for (int i = 0; i < n_needle; ++i) {
      if (dict_is_missing(d_needles, i)) {
        p_out[i] = NA_INTEGER;
        continue;
      }
      
      uint32_t hash = dict_hash_with(d, d_needles, i);
      if (d->key[hash] == DICT_EMPTY) {
        // TODO: Return `no_match` instead
        p_out[i] = NA_INTEGER;
      } else {
        p_out[i] = d->key[hash] + 1;
      }
    }
  } else {
    for (int i = 0; i < n_needle; ++i) {
      uint32_t hash = dict_hash_with(d, d_needles, i);
      if (d->key[hash] == DICT_EMPTY) {
        // TODO: Return `no_match` instead
        p_out[i] = NA_INTEGER;
      } else {
        p_out[i] = d->key[hash] + 1;
      }
    }
  }
```

Or if that is too ugly:

```c
if (propagate) {
    vec_match_loop_propagate(out, d, d_needles, n_needle)
  } else {
    vec_match_loop(out, d, d_needles, n_needle)
  }
```

Reply:
I like the second version. The higher level we get, the less it makes sense to unbranch loops though, because of the trade off with code duplication.